components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt

```diff
@@ -263,13 +266,148 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
                     "${connectParams.gitpodHost}: ${connectParams.workspaceId}: failed to process workspace updates:",
                     t
                 )
-                errorMessage.text = " failed to process workspace updates ${t.message}"
+                setErrorMessage("failed to process workspace updates ${t.message}")
             }
         }
 
         return GitpodConnectionHandle(connectionLifetime, connectionPanel, connectParams);
     }
 
+    private suspend fun resolveJoinLink(
+        ideUrl: URL,
+        ownerToken: String,
+        connectParams: ConnectParams
+    ): String? {
+        var resolveJoinLinkUrl = "https://24000-${ideUrl.host}/joinLink"
+        if (!connectParams.backendPort.isNullOrBlank()) {
+            resolveJoinLinkUrl += "?backendPort=${connectParams.backendPort}"
+        }
+        return fetchWS(resolveJoinLinkUrl, connectParams, ownerToken)
+    }
+
+    private fun resolveCredentials(
```

Comment:
@iQQBot relevant functions are `resolveCredentials`, `resolveHostKeys` and most interesting is `acceptHostKey`